components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/dao/impl/AuthenticatorManagementDAOImpl.java

```diff
@@ -0,0 +1,324 @@
+import org.wso2.carbon.identity.application.common.exception.AuthenticatorMgtException;
+import org.wso2.carbon.identity.application.common.exception.AuthenticatorMgtServerException;
+import org.wso2.carbon.identity.application.common.model.Property;
+import org.wso2.carbon.identity.application.common.model.UserDefinedLocalAuthenticatorConfig;
+import org.wso2.carbon.identity.base.AuthenticatorPropertyConstants.AuthenticationType;
+import org.wso2.carbon.identity.base.AuthenticatorPropertyConstants.DefinedByType;
+import org.wso2.carbon.identity.core.util.IdentityDatabaseUtil;
+import org.wso2.carbon.identity.core.util.IdentityTenantUtil;
+
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * This class implements the AuthenticatorManagementDAO interface which perform CRUD operation on database.
+ */
+public class AuthenticatorManagementDAOImpl implements AuthenticatorManagementDAO {
+
+    private static final Log LOG = LogFactory.getLog(AuthenticatorManagementDAOImpl.class);
+    public static final String IS_TRUE_VALUE = "1";
+    public static final String IS_FALSE_VALUE = "0";
+    public static final String LOCAL_IDP_NAME = "LOCAL";
+
+    @Override
+    public UserDefinedLocalAuthenticatorConfig addUserDefinedLocalAuthenticator(
+            UserDefinedLocalAuthenticatorConfig authenticatorConfig,
+            int tenantId, AuthenticationType type) throws AuthenticatorMgtException {
+
+        Connection dbConnection = IdentityDatabaseUtil.getDBConnection(true);
+
+        try (NamedPreparedStatement statement = new NamedPreparedStatement(dbConnection, Query.ADD_AUTHENTICATOR_SQL)) {
+            statement.setString(Column.NAME, authenticatorConfig.getName());
+            statement.setString(Column.DISPLAY_NAME, authenticatorConfig.getDisplayName());
+            statement.setString(Column.DEFINED_BY, authenticatorConfig.getDefinedByType().toString());
+            statement.setString(Column.AUTHENTICATION_TYPE, type.toString());
+            statement.setInt(Column.IS_ENABLED, authenticatorConfig.isEnabled() ? 1 : 0);
+            statement.setString(Column.IDP_NAME, LOCAL_IDP_NAME);
+            statement.setInt(Column.TENANT_ID, tenantId);
+            statement.executeUpdate();
+
+            if (authenticatorConfig.getProperties() != null) {
+
+                int authenticatorConfigID = getAuthenticatorIdentifier(dbConnection, authenticatorConfig.getName(),
+                        tenantId);
+                addAuthenticatorProperties(dbConnection, authenticatorConfigID, authenticatorConfig.getProperties(),
+                        tenantId);
+            }
+            IdentityDatabaseUtil.commitTransaction(dbConnection);
+
+            return getUserDefinedLocalAuthenticatorByName(dbConnection, authenticatorConfig.getName(), tenantId);
+        } catch (SQLException | AuthenticatorMgtException e) {
```

Comment:
From where does this AuthenticatorMgtException gets thrown ?

Looking at this I feel that particular logic should go to a facade